src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -416,11 +416,7 @@ async def run_setup_py(req: RunSetupPyRequest, setuptools: Setuptools) -> RunSet
             output_filename="setuptools.pex",
             internal_only=True,
             requirements=PexRequirements(setuptools.all_requirements),
-            interpreter_constraints=(
-                req.interpreter_constraints
-                if setuptools.options.is_default("interpreter_constraints")
-                else PexInterpreterConstraints(setuptools.interpreter_constraints)
-            ),
+            interpreter_constraints=req.interpreter_constraints,
```

Comment:
I think this is sound, but note that sometimes users may want to override the constraints used. For example, our workflow in https://github.com/pantsbuild/pants/pull/10951. As mentioned in the PR description, it can matter which interpreter is used.

With this PR's change, the user will need to change the interpreter_constraints for their specific target, rather than being able to force override `[setuptools].interpreter_constraints`. This is less flexible. Although, I think that is consistent with the rest of Pants— and it removes magic of Pants auto-setting things if not configured.

Reply:
The original setup makes no sense. The interpreter constraint for setup tools _must_ be the interpreter constraint of python_distribution. A global value set in the setuptools subsystem can only be right if it matches all of those in the repo and generally will be wrong.

Reply:
Okay, that makes a lot of sense to me. Thanks.

We'll need to restore the old behavior for the sake of the deprecation warning - but sounds good for once the deprecation is over.

Reply:
In fact - is there any reason to put constraints on setuptools and wheel by default in the Setuptools subsystem? It seems we should let them freely float so that the latest compatible versions for the interpreter selected for the `python_distribution` in question are used. Today the version range for setuptools breaks packaging python2.7 `python_distributions` out of the box ,needlessly fwict, for example.

Reply:
Generally, we started pinning things about a year ago iirc to give more stability to our users. We don't eagerly encourage users to pin things like `[setuptools].version`, so this is meant to avoid their build breaking overnight.

You're right that it breaks Py2 users. It's a tradeoff, and we've been biasing towards optimizing for the Py3-only experience, so long as we still have an escape hatch for Py2 users. For example, we use Pytest 6 by default, which breaks Py2 users by default.

Reply:
Ditto breaking python3.10 out of the box.

Reply:
I'll post a separate PR for the un-needed constraints issue.

Reply:
OK - The breaking is needless but this is all a bit messy. If the `--version` - which is an odd name - requirement was a list we could supply a disjoint list of pinned requirements by default by leveraging environment markers in the requirement strings.

Reply:
Pinning _and_ not breaking folks out of the box seems like a nicer path to me. I'll see what that takes and spin up a PR if its not too bad.